libpysal/weights/raster.py

```diff
@@ -0,0 +1,297 @@
+from .set_operations import w_subset
+from .util import lat2W, lat2SW
+from .weights import WSP
+from scipy import sparse
+import numpy as np
+from warnings import warn
+
+try:
+    from numba import njit, jit
+except (ImportError, ModuleNotFoundError):
+
+    def jit(*dec_args, **dec_kwargs):
+        """
+        decorator mimicking numba.jit
+        """
+
+        def intercepted_function(f, *f_args, **f_kwargs):
+            return f
+
+        return intercepted_function
+
+    njit = jit
+
+try:
+    from xarray import DataArray
+except ImportError:
+    raise ImportError(
+        "xarray must be installed to use this functionality")
+
+__all__ = ['da_checker', 'da2W', 'da2WSP', 'w2da', 'da2W2', 'da2WSP2']
+
+
+def da2W(da, criterion, **kwargs):
+    """
+    Create a W object from rasters(xarray.DataArray)
+    Parameters
+    ----------
+    da         : xarray.DataArray
+                 raster file accessed using xarray.open_rasterio method
+    criterion  : {"rook", "queen"}
+                 option for which kind of contiguity to build
+    Returns
+    -------
+    w    : libpysal.weights.W
+           instance of spatial weights class W
+    """
+    if criterion != 'rook':
+        rook = False
+    else:
+        rook = True
+    w = lat2W(*da[0].shape, rook, **kwargs)
+    ser = da.to_series()
+    id_order = np.where(ser != da.nodatavals[0])[0]
+    w = w_subset(w, id_order)
+    ser = ser[ser != da.nodatavals[0]]
+    w.coords = ser.index
+    attrs = da.attrs
+    attrs["dims"] = da.dims
+    attrs["shape"] = da.shape
+    w.attrs = attrs
+    return w
+
+
+def da2W2(da, criterion, **kwargs):
```

Comment:
earlier `W` builder was using `lat2W`, i noticed that building weights from `lat2W` was slower than first building `WSP `object using `lat2SW` and then converting it to `W`